[1단계 방탈출 결제 / 배포] 띠용(주용은) 미션 제출합니다 🚀🚀🚀

jye326 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
가독성이 좋아 보이지 않습니다. 개선방안이 있으면 좋겠네요

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

jye326 committed:
initial commit : 방탈출 예약 대기 코드 마이그레이션

jye326 committed:
feat(resources) : 토스 결제 위젯 연동 및 예약 로직 수정
- reservation.html: 토스 결제 위젯 스크립트 추가 및 UI 구성 변경
- toss-style.css: 결제창 스타일 정의
- user-reservation.js: 토스 결제 위젯 초기화 및 예약 로직 수정
  - 예약 생성 시 결제 요청 추가
  - fetchReservationPayment 함수로 결제 데이터 처리 로직 구현

jye326 committed:
refactor(js) : 불필요한 주석 제거 및 문서 추가
- user-reservation.js 내 완료된 TODO 주석 제거
- 예약 대기 및 결제 플로우 관련 README.md 추가

jye326 committed:
feat(api): 결제 관련 예약 생성 로직 추가
- ReservationService에 `addAndGetWithoutPayment` 메서드를 사용하도록 변경
- ReservationRequest에 결제 정보를 포함한 필드(paymentKey, orderId, amount, paymentType) 추가

jye326 committed:
feat(service) : 예약 생성 시 결제 승인 로직 추가
- TossPaymentClient를 활용한 결제 승인 로직 추가
- addAndGet 메서드를 addAndGetWithoutPayment와 addAndGet로 분리
- User, ReservationTime, Theme 조회를 위한 메서드 각각 추가 (getUser, getReservationTime, getTheme)
- 관련 테스트 코드 수정

jye326 committed:
feat(payment) : Toss Payments 클라이언트 및 결제 승인 처리 로직 추가
- ClientConfig를 통해 TossPaymentClient 생성 빈 추가
- TossPayments API와 상호작용하는 TossPaymentClient 로직 개발
- 결제 승인 관련 DTO (PaymentApproveDto 및 PaymentApproveResponseDto) 추가

jye326 committed:
feat(exception): 결제 승인 실패에 대한 커스텀 예외 및 에러 코드 추가

jye326 committed:
feat(exception): PaymentApproveException 예외 처리 추가
- PaymentApproveException에 @Getter 적용하여 코드 간소화
- GlobalExceptionHandler에 PaymentApproveException 핸들러 추가

jye326 committed:
refactor(payment): 결제 금액 및 실패 메시지 처리 개선
- 비정상적인 secretKey 값 수정
- 결제 금액(paymentAmount) 테스트 값 조정
- 결제 실패 시 오류 메시지가 alert에 표시되도록 변경

jye326 committed:
feat(config, widget): 결제 금액 및 키 변경
- application.yml에 결제 secret-key 정보 추가
- user-reservation.js에서 결제 금액(paymentAmount) 200,000으로 변경

jye326 committed:
feat(payment): TossPaymentClient refactor 및 secretKey 주입 기능 추가
- ClientConfig에 `@Value`로 secretKey 주입 로직 추가
- TossPaymentClient에 secretKey DI(Dependency Injection) 방식으로 변경
- approvePayment 메서드 내 기존 하드코딩된 secretKey 제거
- RestClient 호출 시 오류 처리 방식 개선 (onStatus 메서드 로직 보완)

jye326 committed:
fix(exception): PaymentApproveException 상속 및 로깅 메시지 수정
- RuntimeException 상속 시 `super` 호출 추가
- 로깅 메시지에서 AuthorizationException을 PaymentApproveException으로 변경

jye326 committed:
test(TossPaymentClient) : Toss Payment 클라이언트 테스트 추가
- 결제 시간이 만료된 경우 예외 처리 테스트
- 잘못된 키 요청 시 예외 처리 테스트

jye326 committed:
refactor(reservation) : 결제 승인 로직 리팩토링 및 불필요한 DTO 제거

- PaymentApproveResponseDto 삭제
- ReservationService 내 `addAndGetWithoutPayment` 메서드 삭제 및 기존 로직 합병
- TossPaymentClient 내 `approvePayment` 메서드 반환값 제거 및 무의미한 DTO 응답 삭제

jye326 committed:
feat(service): 결제 승인 및 예약 생성 테스트 추가
- 결제 승인 요청을 포함한 신규 예약 생성 테스트 케이스 추가
- 대기를 결제 없이 생성하는 테스트 케이스 추가
- TossPaymentClient mock 및 관련 검증 로직 추가

jye326 committed:
feat(payment): TossPaymentClient DI 리팩토링 및 secretKey 관리 방식 개선

- TossPaymentClient 내 secretKey 주입 방식 @Value 애노테이션으로 변경
- ClientConfig 내 TossPaymentClient 생성 로직 단순화 및 RestClient 빈으로 분리
- application.yml에서 결제 secretKey 관리 및 테스트용 설정 추가
- TossPaymentClient의 비효율적인 의존성과 구성 로직 정리

jye326 committed:
test(payment): TossPaymentClient 테스트 개선 및 새로운 케이스 추가

- MockRestServiceServer를 활용한 정상 결제 승인 요청 테스트 추가
- @RestClientTest 애노테이션 사용으로 테스트 환경 설정 개선
- 잘못된 secretKey에 대한 예외 처리 테스트 로직 보완
- 불필요한 하드코딩된 secretKey 제거 및 테스트 데이터 개선
- 연결 시간 및 응답 시간 관련 예외 테스트 주석 추가

EastHShin commented on src/main/java/roomescape/business/service/ReservationService.java:
```diff
@@ -38,26 +40,68 @@ public class ReservationService {
+    public ReservationDto addAndGet(final LocalDate date, final String timeIdValue,
+                                    final String themeIdValue,
+                                    final String userIdValue, final ReservationStatus reservationStatus,
+                                    final String paymentKey, final String orderId, final Long amount) {
```
coderabbit 이 제안한 것과 같이, 파라미터가 너무 많네요 :)
DTO를 고려해보면 좋겠습니다.

jye326 replied:
1. Request DTO 를 그대로  Service Layer에 넘기기
    1. Layer 간 책임 분리를 위해 새로운 DTO 판다.
2. Service Layer DTO 새로 파기
    1. Request DTO와 모든 필드 값이 같은 DTO 파기
    2. **Request DTO를 Reservation과 Payment관련 정보 DTO로 나누기**
        1. 가독성, 활용도 이점
---
수정하였습니다!

EastHShin commented on src/main/java/roomescape/infrastructure/payment/TossPaymentClient.java:
```diff
@@ -0,0 +1,44 @@
+@Component
+public class TossPaymentClient {
+
+    private final RestClient restClient;
```
HttpClient 로 RestClient 를 선택해주신 기준은 무엇이었나요?

jye326 replied:
HTTP Client를 처음 사용하게 되어 일단 학습 테스트에서 제공하는 RestClient와 RestTemplate 중에 결정하기로 하였습니다. 체감 상 지원하는 기능은 비슷하나 RestClient의 체이닝이 가독성이 더 뛰어나고, RestClient가 내부적으로 RestTeamplate을 사용하는 최신 라이브러리이므로 RestTempalte의 단점을 보완하기 위해 등장하지 않았을까 추측하여 RestClient로 결정했습니다. 미션 진행 중 메서드 체이닝의 직관성으로 인한 편의를 체감하여 다시 결정한다고 해도 RestClient를 선택할 것 같습니다!

EastHShin commented on src/main/java/roomescape/infrastructure/payment/ClientConfig.java:
```diff
@@ -0,0 +1,16 @@
+    public RestClient restClient(RestClient.Builder builder) {
+        return builder
+                .baseUrl("https://api.tosspayments.com")
+                .build();
```
외부 결제 API 문제로 응답이 굉장히 느리게 오면 어떻게 될까요? 현재의 설정에서 어떤 문제가 발생할지 정리해주시고, 개선 방안을 도출해보시면 좋을 것 같아요 :)

jye326 replied:
응답 시간이 너무 길어지면 해당 클라이언트가 무제한 블로킹 되는 문제가 발생합니다. 따라서 requestFactory를 활용하여 연결 시간 및 수신 시간 제한을 설정했습니다. 제한 시간을 초과하는 연결 또는 응답의 경우 ResourceAccessException 이 발생하므로 해당 예외에 대한 처리를 고려할 수 있습니다!!

EastHShin commented on src/test/java/roomescape/infrastructure/payment/TossPaymentClientTest.java:
```diff
@@ -0,0 +1,79 @@
+                .hasMessage("인증되지 않은 시크릿 키 혹은 클라이언트 키 입니다.");
+    }
+
+    // 연결 시간 예외 테스트
+    // 응답 시간 예외 테스트
```
이 테스트는 TODO 인 건가요?

jye326 replied:
넵. 그런데 연결 시간 및 응답 시간 제한 설정에 대해 테스트할 마땅한 방법을 찾지 못해 삭제하였습니다 😢

EastHShin commented on src/main/java/roomescape/business/service/ReservationService.java:
```diff
@@ -38,26 +40,68 @@ public class ReservationService {
+        return ReservationDto.fromEntity(reservation);
+    }
+
+    public ReservationDto addAndGetWithoutPayment(final LocalDate date, final String timeIdValue,
```
예약 대기는 결제 없이 가능한 상황이군요.
그렇다면 어드민에서 예약을 취소해 예약 대기가 예약으로 전환되었을 때 결제는 어떻게 되는걸까요?
현재 상황에서는 결제 없이 예약 상태로 넘어가네요 :)

jye326 replied:
대기자의 결제 방식은 구현 시 가장 고민했던 부분입니다. 

말씀하신 대로 현재 예약 대기자는 결제 없이 예약 상태로 전환됩니다. 대기자가 결제를 진행하는 시점에 대해서 생각해보았고, 세 가지의 방법이 떠올랐습니다.

1. 예약처럼 대기도 선결제하기
    - 방탈출이라는 도메인 특성 상 대기가 예약으로 전환되는 경우는 흔치 않은 경우이며, 예약으로 전환되지 않고 취소되는 경우가 일반적이라고 생각했습니다. 따라서 취소된 대기들에 대해 각각 환불을 해줘야 하는 과정이 더 빈번하게 발생할 것이고, 이러한 과정이 불필요한 리소스 낭비라고 생각하여 기각했습니다.
2. 관리자의 수동 승인으로 예약 전환 방식을 변경한 뒤, 해당 승인 시에 결제 요청.
    - 대기 1번의 예약 전환은 자동으로 이루어져야 한다고 생각합니다. 결제를 위해 예약 전환 방식을 변경하고 싶지 않았고, 결제를 위해 대기자의 예약 전환을 일일히 수동으로 승인하는 것이 불필요하게 느껴져 기각하였습니다.
3. 기존 방식을 유지(예약 취소 발생 시, 대기 1번은 자동으로 예약 처리)
    - 결제 여부는 Payment 엔티티를 통해 관리되고 있기 때문에 예약에 대한 결제 여부를 쉽게 확인할 수 있습니다. 따라서 현재 상태를 유지하되 실제로 고객이 예약한 시간에 찾아왔을 경우 결제 정보를 확인하여 필요한 경우 현장 결제를 유도하는 것이 가장 합리적이라고 생각하여 채택하였습니다.
    
따라서 이러한 형태로 구현하였습니다.

EastHShin replied:
그러한 의도를 가지고 해주셨군요ㅎㅎ
어떠한 흐름으로 예약대기에 대한 처리가 진행되기를 의도하셨는지 궁금해 남긴 코멘트였습니다~

EastHShin reviewed (changes requested):
안녕하세요 띠용~

이번 미션 함께하게 된 이스트입니다.
몇가지 코멘트 남겨두었으니 확인 부탁드려요 :)

jye326 committed:
feat(config): RestClient 연결 및 응답 타임아웃 설정 추가

- ClientConfig에 SimpleClientHttpRequestFactory 설정 추가
- RestClient 요청에 connectTimeout 및 readTimeout 값 적용 (10초로 설정)
- TossPaymentClient 테스트 코드에서 불필요한 주석 제거

jye326 committed:
feat(reservation): 예약 생성 로직에 DTO 추출 및 결제 관리 리팩토링

- `ReservationSpecDto` 및 `PaymentApproveDto` 추가로 메서드 파라미터 단순화
- `addAndGet` 및 `addAndGetWithoutPayment` 로직 통합
- `ReservationSpecDto` 사용으로 예약 생성 관련 데이터 캡슐화
- 결제 승인 관련 로직 개선 및 내구성 강화
- 테스트 코드에서 DTO 변경에 따른 수정 및 리팩토링 적용

jye326 committed:
refactor(config): ClientConfig 패키지 구조 변경 및 테스트 수정

- `ClientConfig`를 `roomescape.infrastructure.payment.config` 패키지로 이동
- `TossPaymentClientTest`에서 새로운 패키지 구조 반영
- 테스트 환경 설정(`@RestClientTest`)에서 `ClientConfig` 추가

EastHShin commented on src/main/java/roomescape/infrastructure/payment/config/ClientConfig.java:
```diff
@@ -0,0 +1,22 @@
+    public RestClient restClient(RestClient.Builder builder) {
+        SimpleClientHttpRequestFactory requestFactory = new SimpleClientHttpRequestFactory();
+        requestFactory.setConnectTimeout(10000);
+        requestFactory.setReadTimeout(10000);
```
connectTimeout, readTimeout의 차이는 무엇이고, 10초로 선택해주신 기준은 무엇이었나요?

jye326 replied:
connectionTImeout은 서버와의 TCP 연결을 맺는데 걸리는 최대 시간이고, readTimeout은 응답을 기다리는 최대 시간입니다.

10초는 별 고민 없이 임의로 설정한 값입니다. 연결 타임 아웃은 일반적인 설정 시간을 참고하여 3초로 변경하였고, 응답 타임 아웃은 [토스 문서](https://docs.tosspayments.com/resources/glossary/timeout)를 참고하여 권장 시간 범위의 중간 값인 45초로 설정하였습니다.

EastHShin replied:
위 문서에서 권장한 것은 신용카드 결제와 같이 오래 걸리는 API에 대해 30\~60 초로 권장한 것 같네요.
방탈출 어플리케이션 서버에서 사용하는 confirm API 에 대해서는 그렇게 긴 시간을 설정할 필요는 없을 것 같습니다.
timeout을 길게 설정했을 때, 외부 API가 장애상황이고 사용자 요청이 몰려든다면 어떤 문제가 생길까요?

EastHShin commented on src/test/java/roomescape/infrastructure/payment/TossPaymentClientTest.java:
```diff
@@ -0,0 +1,77 @@
+
+    // 올바른 클라이언트 연결 테스트
+    @Test
+    void 정상_결제_승인_요청() {
```
테스트가 실패하네요!

jye326 replied:
factory를 설정한 RestClient는 Mocking이 안되는 문제가 있었습니다! 
RestClientCustomizer 를 활용하여 test가 Builder를 가로챌 수 있도록 수정하였습니다!

EastHShin commented on src/test/java/roomescape/infrastructure/payment/TossPaymentClientTest.java:
```diff
@@ -0,0 +1,77 @@
+
+    @Test
+    void 잘못된_키로_요청을_보내면_예외가_발생한다() {
+        // given
+        tossPaymentClient = new TossPaymentClient(
+                RestClient.builder()
+                        .baseUrl("https://api.tosspayments.com").build()
+                ,
+                new Jackson2ObjectMapperBuilder().createXmlMapper(false).build(),
+                "invalid" + secretKey
+        );
+        PaymentApproveDto paymentApproveDto = new PaymentApproveDto("paymentKey", "1", 1000L);
+        // when
+        Assertions.assertThatThrownBy(() -> tossPaymentClient.approvePayment(paymentApproveDto))
+                .isInstanceOf(PaymentApproveException.class)
+                .hasMessage("인증되지 않은 시크릿 키 혹은 클라이언트 키 입니다.");
+    }
```
실제 외부 API를 호출하는 테스트인가요?
만약 외부 API가 정상적으로 제공되지 않는 상태라면 테스트가 실패할텐데, 이에 대해서는 어떻게 생각하시나요?

jye326 replied:
외부 API가 정상적으로 제공되지 않는 경우에 위 테스트는 실패하는 것이 맞다고 생각합니다. 애플리케이션이 외부 API로 승인 요청을 보내는 것에 대한 내부 테스트는 Mocking을 활용해 외부 API 상태와 독립적으로 진행됩니다. 따라서 외부 API에 문제가 발생한다면 Mocking 테스트는 성공하고, 나머지 실제 호출 테스트는 실패할 것이므로 애플리케이션 내부의 문제가 아니라 외부 API의 문제임을 파악하기 용이할 것입니다.

따라서 실제 API를 호출하도록 테스트를 설계했고, 토스에서 제공하는 테스트용 url이기 때문에 요청을 여러번 보내도 문제될 것이 없다고 판단했습니다!

EastHShin replied:
테스트는 기본적으로 멱등성을 가지고 있어야 한다 생각합니다. 방탈출 어플리케이션 서버가 외부 요인으로 인해 테스트가 실패하면 안되고, 항상 같은 결과를 반환해야 합니다.
테스트 코드는 외부 API를 테스트하고자 함이 아니라, 비즈니스 로직을 테스트하고자 함이 목적이 되어야 한다 생각해요.
물론, 외부 API에 대한 호출 및 응답 변환과정에서의 로직을 테스트하기 위해 인프라 계층 단위테스트는 필요하지만, 이 또한 외부 API 호출과는 독립적으로 수행되어야 한다 생각해요.

EastHShin reviewed (changes requested):
안녕하세요 띠용

피드백 반영 잘 해주셨네요~
추가 코멘트 남겨두었으니 확인 부탁드려요 :)

jye326 committed:
feat(payment): RestClient 설정 분리 및 TossPaymentClient 리팩토링

- ClientConfig에서 RestClient 생성 로직과 RestClientCustomizer 분리
- TossPaymentClient의 RestClient DI 방식을 Builder로 유연하게 변경
- RestClient 연결 타임아웃(connectTimeout)과 응답 타임아웃(readTimeout) 축소 (10초 → 5초)
- TossPaymentClient 테스트에서 MockRestServiceServer 설정 방식 개선
- TossPaymentClient 생성자에서 불필요한 @Component 애노테이션 제거

jye326 committed:
feat(config): RestClient 연결 및 응답 타임아웃 수정

- ClientConfig에서 connectTimeout 값을 5초 → 3초로 감소
- readTimeout 값을 5초 → 45초로 증가하여 긴 응답 대기 처리 가능하도록 수정

EastHShin reviewed (approved):
안녕하세요 띠용

이번 단계 요구사항은 모두 충족해 마무리하도록 하겠습니다.
코멘트 답변 남겨두었으니 확인해주시고, 다음 단계에서 뵙겠습니다 :)
